contracts/src/zaps/UniV3Zap.sol

```diff
@@ -0,0 +1,937 @@
+import { UniV3Path } from "../libraries/UniV3Path.sol";
+
+/// @title UniV3Zap
+/// @author DELV
+/// @notice A zap contract that uses Uniswap v3 to execute swaps before or after
+///         interacting with Hyperdrive.
+/// @custom:disclaimer The language used in this code is for coding convenience
+///                    only, and is not intended to, and does not, have any
+///                    particular legal or regulatory significance.
+contract UniV3Zap is IUniV3Zap, ReentrancyGuard {
+    using FixedPointMath for uint256;
+    using SafeERC20 for ERC20;
+    using UniV3Path for bytes;
+
+    /// @dev We can assume that almost all Hyperdrive deployments have the
+    ///      `convertToBase` and `convertToShares` functions, but there is
+    ///      one legacy sDAI pool that was deployed before these functions
+    ///      were written. We explicitly special case conversions for this
+    ///      pool.
+    address internal constant LEGACY_SDAI_HYPERDRIVE =
+        address(0x324395D5d835F84a02A75Aa26814f6fD22F25698);
+
+    /// @notice We can assume that almost all Hyperdrive deployments have the
+    ///         `convertToBase` and `convertToShares` functions, but there is
+    ///         one legacy stETH pool that was deployed before these functions
+    ///         were written. We explicitly special case conversions for this
+    ///         pool.
+    address internal constant LEGACY_STETH_HYPERDRIVE =
+        address(0xd7e470043241C10970953Bd8374ee6238e77D735);
+
+    /// @notice The Uniswap swap router.
+    ISwapRouter public immutable swapRouter;
+
+    /// @notice The wrapped ether address.
+    IWETH public immutable weth;
+
+    /// @notice The name of this zap.
+    string public name;
+
+    /// @notice The kind of this zap.
+    string public constant kind = UNI_V3_ZAP_KIND;
+
+    /// @notice The version of this zap.
+    string public constant version = VERSION;
+
+    /// @notice Instantiates the zap contract.
+    /// @param _name The name of this zap contract.
+    /// @param _swapRouter The uniswap swap router.
+    /// @param _weth The wrapped ether address.
+    constructor(string memory _name, ISwapRouter _swapRouter, IWETH _weth) {
+        name = _name;
+        swapRouter = _swapRouter;
+        weth = _weth;
+    }
+
+    /// Receive ///
+
+    /// @notice Allows ETH to be received within the context of a zap.
+    /// @dev This fails if it isn't called within the context of a zap to reduce
+    ///      the likelihood of users sending ether to this contract accidentally.
+    receive() external payable {
+        // Ensures that the zap is receiving ether inside the context of another
+        // call. This means that the ether transfer should be the result of a
+        // Hyperdrive trade or unwrapping WETH.
+        if (!_reentrancyGuardEntered()) {
+            revert InvalidTransfer();
+        }
+    }
+
+    /// LPs ///
+
+    /// @notice Executes a swap on Uniswap and uses the proceeds to add
```

Comment:
could you just inherit this comment from the interface?

Reply:
I'm glad that you bring this up. We definitely can, but I personally don't love that kind of inheritance because it moves the documentation away from the code that it is documenting. It is definitely good for maintainability and documentation duplication, but it's bad for readability IMO.

I think we need to collectively agree on what tradeoff we want to make. This is my personal preference, but there may be strong arguments in the other direction that would change my mind.